Basic AcroForms input controls rendering

xlc opened the pull request:
There is already an example implementation of AcroForms in the examples folder. This PR aim to do it properly.

This PR only provide basic implementation of text input fields. Other types of fields (buttons, choice) will be implemented in later PRs.

Issues & Questions:
- Printing: Currently only buttons are rendered in canvas and other form elements are not printed.
  - We need to either:
    - Render the DOMs before printing?
    - Or render all of them in canvas?
- Appearance: Need a way to process appearances and apply them to the DOMs.
- How to handle user input values?
  - One way is not expose any JS API. If you need to listen for value change, just do `$('.annotationLayer .widgetControl').on('change', changeHandler);`
  - Or add some method on `PDFPageView` to read/write/listen values.

timvandermeij commented:
/botio-linux preview

xlc commented:
Can anyone point me how to parse/evaluate "a sequence of valid page-content graphics or text state operators that define such properties as the field’s text size and color" ?
e.g.  `/TimesNewRomanPSMT 12 Tf 0 0 1 rg`
It is the way to set text input's background color and font. I can't find any way to use existing code to process it.

timvandermeij commented:
Let's keep things simple and avoid parsing appearance streams for now. The reason I say this is that this PR has a greater chance of being merged if it is smaller and therefore easier to review and test. We need to split the work here into several PRs and iterate on the issue. For instance, let this PR just be the basic AcroForms framework with basic fields, then a next PR for the other fields, and a next PR for annotation appearance streams. The latter is also affecting my work on the annotation layer, so we need to find out the best way to implement that according to the specification.

xlc commented:
@timvandermeij Thanks for your suggestion. Can you review the changes?

fbender commented:
@xlc would you mind splitting up this PR into smaller ones? 
@timvandermeij How far along have you come with the annotation layer? What changes have already been merged that obsolete (part of) this PR?

Also see kind-off tracking bug #1459

xlc commented:
@fbender How should I splitting this? This PR only implements text fields. I have more controls implemented in my other branch which depends on this. I will PR it when this is merged in.

timvandermeij commented:
Fortunately most of my PR is already merged using smaller PRs (see https://github.com/mozilla/pdf.js/pull/5218#issuecomment-155598653 for a complete explanation) and I hope to do the rest soon such that that PR is done. It should not block this PR too much, but we need to have more structure first before adding new annotation features (AcroForms from here, but also the video annotation PR). Part of my PR is the handling of flags in the base class such that subclasses do not need to implement that too (as I see some flag handling in this PR). I also feel that the elements processed here should be in a class that extends the Annotation base class as AcroForm elements are Widgets if I'm not mistaken. We should unify the handling of those somewhat better here.

fbender commented:
> How should I splitting this?

See e.g. @timvandermeij's answer:

> For instance, let this PR just be the basic AcroForms framework with basic fields, then a next PR for the other fields, and a next PR for annotation appearance streams.

This needs to be adapted to the new stuff implemented in #5218 first, of course. See e.g.

> but we need to have more structure first before adding new annotation features […]. Part of my PR is the handling of flags in the base class such that subclasses do not need to implement that too […]. I also feel that the elements processed here should be in a class that extends the Annotation base class as AcroForm elements are Widgets if I'm not mistaken. We should unify the handling of those somewhat better here.

This somehow seems to contradict this statement:

> It should not block this PR too much

@timvandermeij can you please clarify which changes you expect to be blocking this work? Would be nice if there are (merged or still unmerged) PRs (or implementation concepts/drafts) you can link to that will help @xlc in refactoring his code.

timvandermeij commented:
I meant to say that there are no real dependencies on #5218 for this PR, but that this PR does need a bit more structure and clean-up. Examples:

- It needs to be rebased as some of the used constructs are not available in master anymore.
- The preference name should have "disable" in it instead of "enable" to be in line with all other preferences.
-  I'm not a fan of adding the `getInteractiveHtmlElement` function as it is kind of duplicating logic in `getHtmlElement`.
- We should split up the Widget annotation handling into text widgets and form widgets. They are all Widget annotations, but serve a very different purpose. We already have `getHtmlElementForTextWidgetAnnotation` for regular text widgets, but `createTextWidgetAnnotation` works on form widgets although the name does not imply that.
- It needs unit tests.

I think the bottom line is that I'm not sure on what the best structure is. In the PDF specification, Adobe tends to label everything that does not belong to the other annotation types a Widget, but there are many kinds of Widgets (static text widgets, form widgets like buttons and fields, et cetera). I don't think intertwining them all in the code is a good idea: there has to be a clear distinction between form widgets and text widgets, for instance. Ideally every type of annotation, but also every subtype like FormWidgets or TextWidgets, are separate classes. Other ideas for a good structure are of course welcome.

xlc commented:
AFAIK, Widget annotations are form widgets. There is no static text widget (there is text annotation and free text annotation). Text Field widget is the interactive form widget you can enter text. So I don't think we need distinction between form widgets and text widgets. (Are there any non-form widgets? I can't find form PDF32000 spec)

`getHtmlElementForTextWidgetAnnotation` is used to display non-interactive text widgets and 
`createTextWidgetAnnotation` is used for interactive text widgets. Both functions works with same widget annotation type but just for different modes.

I can do rebase, refactor code, unit tests, rename preference without too much problem, but I agree we need a good structure first.
Currently in order to implement a new form widget (I have already implemented a few in different branch), I have to modify `annotation.js` to add a new class subclass from `WidgetAnnotation`, read whatever fields and parse them and store fields into data object and then modify `annotation_helpers.js` to implement the logic to render the widget with HTML. Ideally the data object should provide a way to expose all required fields so the logic about a particular widget is in a single place. More importantly, I don't want to modify core pdfjs code in order to implement a feature in viewer. So that other viewer implementations can implement new form widgets or new annotation types without modifying the core pdfs code. (I am basically asking for #5283)

timvandermeij commented:
@xlc I agree with you that we need to think more about how to structure this properly. Fixing up this PR should not be too hard, but it's the structure that is really important in order to extend this also to all other annotation types.

timvandermeij commented on src/core/annotation.js:
```diff
@@ -634,7 +635,7 @@ var LinkAnnotation = (function LinkAnnotationClosure() {
           url = '';
         }
-        // According to ISO 32000-1:2008, section 12.6.4.7, 
+        // According to ISO 32000-1:2008, section 12.6.4.7,
```
This change is no longer necessary when you rebase the patch as it has been fixed already in another commit.

timvandermeij commented on src/core/annotation.js:
```diff
@@ -485,6 +485,7 @@ var WidgetAnnotation = (function WidgetAnnotationClosure() {
     var dict = params.dict;
     var data = this.data;
 
+    data.annotationType = AnnotationType.WIDGET;
```
This change is no longer necessary when you rebase the patch as it has been fixed already in #6757.

timvandermeij commented on src/core/annotation.js:
```diff
@@ -547,8 +548,8 @@ var TextWidgetAnnotation = (function TextWidgetAnnotationClosure() {
     WidgetAnnotation.call(this, params);
 
     this.data.textAlignment = Util.getInheritableProperty(params.dict, 'Q');
-    this.data.annotationType = AnnotationType.WIDGET;
```
This change is no longer necessary when you rebase the patch as it has been fixed already in #6757.

timvandermeij commented on src/display/annotation_helper.js:
```diff
@@ -114,6 +118,90 @@ var AnnotationUtils = (function AnnotationUtilsClosure() {
+
+  function createTextWidgetAnnotation(item, commonObjs) {
+    var isMultiline = isBitSet(item.fieldFlags, 13);
+    // var isPassword = isBitSet(item.fieldFlags, 14);
```
Please remove commented out code if it is not used. It can be added in a follow-up patch when it becomes useful.

timvandermeij commented on src/display/annotation_helper.js:
```diff
@@ -273,6 +361,9 @@ var AnnotationUtils = (function AnnotationUtilsClosure() {
   }
 
   function getHtmlElement(data, objs) {
+    if (!data.hasHtml) {
```
This change should not be necessary. It is already handled in https://github.com/mozilla/pdf.js/blob/master/src/display/annotation_layer.js#L330-L332 just before calling this method.

xlc commented:
@timvandermeij I have rebased and updated this PR. 
Do you want me to change preference `enableInteractiveForms` to `disableInteractiveForms`? I prefer `enableInteractiveForms ` because it is disabled by default, similar to `enableWebGL` and `enableHandToolOnLoad`

timvandermeij commented:
I'm actually wondering if we need to make this a preference at all. Why not just render them? @yurydelendik mentioned on IRC that rendering the fields may just be fine because then people can fill them and print the document (see http://logs.glob.uno/?c=mozilla%23pdfjs&s=1+Dec+2015&e=1+Dec+2015&h=printing#c42367). Admittedly they will not be able to save the PDF, but printing the data should be enough for most purposes.

There are no tests for `annotation_layer.js` yet.

xlc commented:
@timvandermeij That's a good point. Chrome's build in PDF reader does exactly that. But the next problem is that currently PDF.js does not print any DOM elements including form fields. So as I commented before, we need to either
- Render DOMs before printing
- Or render all of the annotations in canvas
- Or any other ideas?

timvandermeij commented:
We need to render the DOMs before printing I think, which is already done for e.g. Text annotations. Try opening the following PDF file and showing a print preview:

[Text_aac_11_01.pdf](https://github.com/mozilla/pdf.js/files/64745/Text_aac_11_01.pdf)

You will see that it gets printed because it has print flags set for the annotations, handled by https://github.com/mozilla/pdf.js/blob/master/src/core/annotation.js#L190-L197. Therefore I think rendering the AcroForms fields should be possible, but I also think it's best to do that in a follow-up PR as that adds more complexity. Let's have this PR first render the fields correctly.

xlc commented:
@timvandermeij In your file, the text annotation is rendered in the canvas so it prints fine. However for [this file](https://drive.google.com/open?id=0B6YhVF9p-y2AYmtDbGxUX2lUQkE), the values for form field will not be printed because they are rendered using DOMs and not available in canvas. We have to update the printing code to also render DOMs as well as canvas. But anyway, lets fix printing in next PR.

yurydelendik commented:
So far the problem is if users types information and viewer destroys the page (because it's out of view) the information loss can happen. (At https://github.com/mozilla/pdf.js/blob/master/examples/acroforms/forms.js I introduced the `formFields` storage for this reason) Can we add something like this to this PR? If this storage object can be serialized, if the future we can pass this data to the render(), so it can generate operator list for the printed canvas.

timvandermeij commented:
Please remove the "remove hasHtml from annotation data" commit. That needs some more thought and causes regressions for unsupported annotation types at this point. It is also not related to this PR.

timvandermeij commented:
This PR also needs a rebase to make the code match with the new annotation code.

xlc committed:
implement interactive text forms

xlc commented on src/display/annotation_layer.js:
```diff
@@ -374,19 +386,82 @@ var WidgetAnnotationElement = (function WidgetAnnotationElementClosure() {
-      this.container.appendChild(content);
-      return this.container;
+      if (!isMultiline && !('fontSize' in this.data)) {
+        // hack to guess font size base on content hight
```
An example file that needs this hack is https://drive.google.com/file/d/0B6YhVF9p-y2AYzRNLTE5Z2dFWHM/view?usp=sharing
Without it some of the fields will be cut because the fonts are too large

timvandermeij replied:
Nit: replace `hight` with `height`.

xlc commented:
Still depends on #6818 or #6786 to remove/fix `hasHtml` so empty text fields can be rendered.
Or @timvandermeij do you want me make another PR to remove `hasHtml`?

timvandermeij commented:
I just added some more review comments. This is now going in the right direction. I'm open to ideas to remove `hasHtml` because I don't like it either, but so far I was not successful with removing it without causing regressions. I think we should remove `hasHtml` in a separate PR, before landing this PR since this PR depends on it.

timvandermeij commented:
This PR needs a rebase now that your other PR landed.

timvandermeij commented on src/display/annotation_layer.js:
```diff
@@ -374,19 +383,88 @@ var WidgetAnnotationElement = (function WidgetAnnotationElementClosure() {
+      if (!isMultiline && !('fontSize' in this.data)) {
+        // hack to guess font size base on content hight
+        // so it display fine on small text field
+        // this need to be removed when we can apply defaultAppearance
```
Nit: `this needs`

timvandermeij commented on src/display/annotation_layer.js:
```diff
@@ -374,19 +383,88 @@ var WidgetAnnotationElement = (function WidgetAnnotationElementClosure() {
+      var isMultiline = this._hasFlag(MULTILINE_BIT);
+
+      var content;
+
```
Nit: remove the newline here.

timvandermeij commented on src/display/annotation_layer.js:
```diff
@@ -374,19 +383,88 @@ var WidgetAnnotationElement = (function WidgetAnnotationElementClosure() {
+
+  Util.inherit(TextWidgetAnnotationElement, WidgetAnnotationElement, {
+    /**
+     * Render the text widget annotation't HTML element.
```
Nit: `annotation's`

timvandermeij commented on src/display/annotation_layer.js:
```diff
@@ -67,7 +67,16 @@ AnnotationElementFactory.prototype =
+          default:
+            warn('Unimplemented Widget annotation type: ' +
+                  parameters.data.fieldType);
+            // Fallback to default one, which does not render anything
```
Nit: add a period at the end of the sentence.

Snuffleupagus replied:
Also, this block seems unnecessarily complicated. Wouldn't e.g. this work just as well:
```diff
-  return new WidgetAnnotationElement(parameters);
+  if (parameters.data.fieldType === 'Tx') {
+    return new TextWidgetAnnotationElement(parameters);
+  }
+  warn('Unimplemented Widget annotation type: ' +
+       parameters.data.fieldType);
+  // Fallback to default one, which does not render anything.
+  return new WidgetAnnotationElement(parameters);

```

timvandermeij replied:
I assume it has been done with a `switch` because the intention is to add more types later, but perhaps until that happens it's better to do it like suggested above.

Snuffleupagus replied:
Perhaps, but even so it seems unnecessarily complex. If that is the case, this should work too:
```diff
-  return new WidgetAnnotationElement(parameters);
+  switch (parameters.data.fieldType) {
+    case  'Tx':
+      return new TextWidgetAnnotationElement(parameters);
+    // case 'whatever':
+    // return new SomeMethodHere(parameters);
+  }
+  warn('Unimplemented Widget annotation type: ' +
+       parameters.data.fieldType);
+  // Fallback to default one, which does not render anything.
+  return new WidgetAnnotationElement(parameters);
```

timvandermeij replied:
Right, I agree.

timvandermeij commented on src/display/annotation_layer.js:
```diff
@@ -374,19 +383,88 @@ var WidgetAnnotationElement = (function WidgetAnnotationElementClosure() {
+     * @memberof WidgetAnnotationElement
+     * @returns {boolean}
+     */
+    _hasFlag: function WidgetAnnotationElement_hasFlag(pos) {
```
Will this be used for multiple classes later on or is it only relevant for the `TextWidgetAnnotation` class? If so, we should move this there.

timvandermeij replied:
Also, let's change `pos` to `position` for readability.

xlc replied:
This will be used by other widget annotation classes including Button and Choice

timvandermeij commented on src/display/annotation_layer.js:
```diff
@@ -374,19 +383,88 @@ var WidgetAnnotationElement = (function WidgetAnnotationElementClosure() {
       content.style.textAlign = ['left', 'center', 'right'][textAlignment];
-      content.style.verticalAlign = 'middle';
-      content.style.display = 'table-cell';
+      if (this.data.maxLen !== null) {
```
Would `if (this.data.maxLen) {` suffice too? I'm not sure if you need to explicitly mention `!== null`.

xlc replied:
0 is a valid maxLen

timvandermeij commented on src/display/annotation_layer.js:
```diff
@@ -374,19 +383,88 @@ var WidgetAnnotationElement = (function WidgetAnnotationElementClosure() {
-      return this.container;
+      if (!isMultiline && !('fontSize' in this.data)) {
+        // hack to guess font size base on content hight
+        // so it display fine on small text field
```
Nit: `so small text fields are rendered correctly`

xlc commented:
@timvandermeij Updated.

timvandermeij commented on src/display/annotation_layer.js:
```diff
@@ -371,19 +376,87 @@ var WidgetAnnotationElement = (function WidgetAnnotationElementClosure() {
-      this.container.appendChild(content);
-      return this.container;
+      if (!isMultiline && !('fontSize' in this.data)) {
+        // hack to guess font size based on content hight
```
Nit: start with a capital H and change "hight"  to "height"

timvandermeij commented:
This is looking much better than before. Could you address the other nit above and then squash the commits into one?

Snuffleupagus commented:
Also, unless I'm overlooking something, isn't https://github.com/mozilla/pdf.js/pull/6172#issuecomment-168014477 still unaddressed?

timvandermeij commented:
Good point, that one has to be addressed too.

xlc commented:
What will be the best way to save the values? Should I create some FormFieldStorage to store the values? Or attach a field somewhere else?

xlc committed:
avoid render form field twice

xlc committed:
add storage service and use it to store form field value

xlc commented on src/core/annotation.js:
```diff
@@ -398,8 +398,17 @@ var Annotation = (function AnnotationClosure() {
+          include = false;
+        }
+      }
+      if (include) {
```
Testing file: https://drive.google.com/file/d/0B6YhVF9p-y2AbS1wY3drMDdGRVE/view?usp=sharing

xlc commented:
@Snuffleupagus @timvandermeij I have implemented a storage solution, what do you think?

xlc commented:
@timvandermeij Can you have a look?

fbender commented:
@xlc The branch needs to be rebased. Care to do that and ping @Snuffleupagus again afterwards?

timvandermeij commented:
Closing as superseded by a series of other PRs that implemented this functionality. Remaining tasks are tracked in #7613.
